relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -1,152 +1,228 @@
+                .inner
+                .iter()
+                .all(|cond| cond.matches(value_provider)),
+            RuleCondition::Or(conditions) => conditions
+                .inner
+                .iter()
+                .any(|cond| cond.matches(value_provider)),
+            RuleCondition::Not(condition) => !condition.inner.matches(value_provider),
+            _ => false,
+        }
     }
 }
 
-impl<'de> Deserialize<'de> for LowerCaseString {
-    fn deserialize<D>(deserializer: D) -> Result<LowerCaseString, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        deserializer.deserialize_string(LowerCaseStringVisitor)
-    }
+/// A sampling rule as it is deserialized from the project configuration  
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SamplingRule {
+    pub condition: RuleCondition,
+    pub sample_rate: f64,
+    #[serde(rename = "type")]
+    pub ty: RuleType,
 }
 
 impl SamplingRule {
-    /// Tests whether a rule matches a trace context
-    fn matches(
-        &self,
-        release: Option<&str>,
-        user_segment: &Option<LowerCaseString>,
-        environment: &Option<LowerCaseString>,
-        project_id: ProjectId,
-        strategy: SamplingStrategy,
-    ) -> bool {
-        // check we are matching the right type of rule
-        if self.strategy != strategy {
-            return false;
-        }
+    fn is_not_supported(&self) -> bool {
+        self.condition.is_not_supported()
+    }
+}
 
-        // match against the environment
-        if !self.environments.is_empty() {
-            match environment {
-                None => return false,
-                Some(ref environment) => {
-                    if !self.environments.contains(environment) {
-                        return false;
-                    }
-                }
-            }
-        }
+/// Trait implemented by providers of fields (Events and Trace Contexts).
+/// The fields will be used by rules to check if they apply.
+trait FieldValueProvider {
+    /// gets the value of a field
+    fn get_value(&self, path: &str) -> FieldValue;
+    /// what type of rule can be applied to this provider
+    fn get_rule_type(&self) -> RuleType;
+}
 
-        // match against the project
-        if !self.project_ids.is_empty() && !self.project_ids.contains(&project_id) {
-            return false;
+impl FieldValueProvider for Event {
+    fn get_value(&self, field_name: &str) -> FieldValue {
+        match field_name {
+            "event.release" => match self.release.as_str() {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s),
+            },
+            "event.environment" => match self.environment.as_str() {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s),
+            },
+            "event.user_segment" => FieldValue::None, // Not available at this time
```

Comment:
Can we make this `event.user` or `event.user.segment`? I am still very much set that we won't be able to introduce the user "segment" concept directly into SDKs, so we will need to provide other user information in place of this.

Same for the trace.

Reply:
I left it as is, I'll change it to something appropriate when we figure out what how are we going to get the data.